client/interface.go

```diff
@@ -64,7 +64,7 @@ type ContainerAPIClient interface {
 	ContainerRemove(ctx context.Context, container string, options types.ContainerRemoveOptions) error
 	ContainerRename(ctx context.Context, container, newContainerName string) error
 	ContainerResize(ctx context.Context, container string, options types.ResizeOptions) error
-	ContainerRestart(ctx context.Context, container string, timeout *time.Duration) error
+	ContainerRestart(ctx context.Context, container string, timeout *time.Duration, options types.ContainerStartOptions) error // TODO create separate types.ContainerRestartOptions?
```

Comment:
A little conflicted on changing the client interface here.
On the one hand we can just go ahead and do it (with a RestartOptions) struct and get it over with but break client code when they update.
On the other we can add a 2nd method with options and leave this one alone, but complicate the interface.

Reply:
Right. Feels like Go's [lack of overloading support](https://golang.org/doc/faq#overloading) creates quite a bit of a burden when it comes to API design. Hard to avoid breaking changes in cases like this. 

I don't have enough context on how breaking client changes have been handled in the past, but changing it one time seems preferable to polluting the interface with a 2nd method to eternity. Thoughts?

Reply:
[Removed](https://github.com/moby/moby/pull/39343/commits/753b59255aa4655d7f646956b53299e27b4a45a2) the `TODO` comment btw. I don't think there's a need for a `types.ContainerRestartOptions` as it would be identical to the existing `types.ContainerStartOptions`.

Reply:
If we're gonna break the interface, should we include `timeout` in the `options` as well? i.e., make the interface something like;

```go
ContainerRestart(ctx context.Context, container string, options types.ContainerRestartStartOptions)
```

Reply:
It's certainly a possibility! If y'all feel strongly about it, I'll be happy to make that change.

I'm a little torn though: `timeout` is for stopping the container, `options types.ContainerStartOptions` for subsequently starting it. So it could be plausible to interpret the parameters along the lines of "try stopping the container for `timeout` seconds, if that succeeds, start it with the following `ContainerStartOptions`".

If, in the future, another parameter gets added to `types.ContainerStartOptions`, it'll immediately work with both `ContainerStart` and `ContainerRestart`.

Thoughts?